Add DTD to Galaxy

ericenns opened the pull request:
This merges DTD upstream into Galaxy.

https://github.com/phac-nml/dynamic-tool-destination

WIP until all tests have been moved over and reviewed fully.

ericenns committed:
Initial DTD to Galaxy work

bgruening commented:
Great poster, great idea, I like it much and I think it is a nice addition to the dynamic tool destinations.
Thanks @ericenns!

ericenns committed:
Added in DTD tests

ericenns committed:
Flake8 fixes

ericenns commented:
I think WIP label can be removed, I have added all our tests in and they pass!

bgruening commented on config/tool_destinations.yml.sample:
```diff
@@ -0,0 +1,302 @@
+# Galaxy's history panel).
+
+
+tools:
```
Should this be commented out?

ericenns replied:
Do you mean should the whole file be commented out? If so yes, is that what is done with out sample config files?

bgruening commented:
@ericenns do you think we should keep the configuration under config or is this something really specific to the dynamic-job-destination and should be next to the job-destination to not pollute the config dir?

ericenns commented:
The config file is specific to dtd, one benefit to keeping it in config is that it gives visibility to it. I guess I could add it as an example to job_conf.xml.sample_advanced

ericenns commented:
Guess I have to remove WIP from the title or galaxybot will keep adding the label back.

jmchilton commented:
@galaxybot test this

jmchilton commented:
Really nifty - I :heart: the tests and focus on validation - this great.

I created https://github.com/galaxyproject/starforge/pull/100 so the wheel for testfixtures can be placed on wheels.galaxyproject.org.

I'm going to open a counter-PR to make the config handling more Galaxy-ish.

jmchilton commented:
The upstream project has a different copyright statement and license than Galaxy and it is featured prominently in README - given that I just wanted to make sure that you are explicitly granting the Galaxy project permission to redistribute this code with our license and copyright.

ericenns commented:
Yes I am granting the Galaxy project permission to redistribute this code with their license and copyright.

jmchilton committed:
Use more traditional Galaxy config handling for DTD.

jmchilton committed:
Ease configuration and add documentation for using DTD syntax for destinations.

Moving rules also means it won't conflict with other peoples custom dynamic destinations and should prevent Galaxy from reloading the Python repeatedly.

jmchilton commented:
Thanks @ericenns!

I have two commits I'd like to discuss but Github isn't letting me open a PR against your repo and I don't understand why. Any chance you can review them, test them, and cherry-pick them into your branch if they look good? 

They are from my branch here - https://github.com/jmchilton/galaxy/tree/add-dtd-to-galaxy.

https://github.com/jmchilton/galaxy/commit/ed48c6c28b72ef63499ff39abcce9171fb684154 Makes the config handling more like traditional Galaxy - navigating the Galaxy config object to find that YAML file.
https://github.com/jmchilton/galaxy/commit/d87957e1a2030752b5617934b911da0855bae3fe Re-imagines how this would be used a bit - it makes this a special "type" of dynamic destination (there were already some others) so details such as the function name don't need to be exposed. It also adds some documentation to job_conf.xml.sample_advanced about how to use it.

Any problems with these changes?

ericenns commented:
Thanks @jmchilton, the changes look good to me. I will merge them and retest.

ericenns committed:
Moved dtd tests into a different folder

ericenns committed:
Updated tool_destinations.yml.sample to include a little more information

ericenns commented:
@jmchilton everything looks good, a revised test location and updated the sample config with a little more information on how to validate your config.

ericenns commented:
@galaxybot test this

ericenns commented:
Hmm looks like there is a problem with running the following command:

```
 bash -c '. .venv/bin/activate; python lib/galaxy/jobs/dynamic_tool_destination.py -c config/tool_destinations.yml.sample'
```

I am looking into it right now

ericenns committed:
Fix dict iteration in dtd default_destination parsing/validation

ericenns commented:
Ok everything should be good to go now

ericenns commented:
@galaxybot test this

jmchilton commented:
Very nice @ericenns - thanks a bunch for this contribution!

jmchilton commented:
And of course I merged this without waiting for my own starforge PR (https://github.com/galaxyproject/starforge/pull/100) :angry:.

bgruening commented:
:tada: This is awesome! Next GCC presentation in Galaxy main! Thanks a lot @ericenns!
